File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter17/advanced/AdvancedDesignPatterns.kt
package com.devlokos.masteringkotlin.chapters.chapter17.advanced

/**
 * Capítulo 17 — Patrones de Diseño.
 * Nivel AVANZADO: Observer — NewsPublisher notifica a suscriptores cuando se publican artículos.
 */
fun main() {
    runAdvancedDesignPatternExamples()
}

typealias Subscriber = (String) -> Unit

class NewsPublisher {
    private val subscribers = mutableListOf<Subscriber>()

    fun subscribe(handler: Subscriber) {
        subscribers.add(handler)
    }

    fun publish(article: String) {
        subscribers.forEach { it(article) }
    }
}

fun runAdvancedDesignPatternExamples() {
    println("========================================")
    println("       OBSERVER PATTERN")
    println("========================================")
    val publisher = NewsPublisher()
    publisher.subscribe { article -> println("  Subscriber A: $article") }
    publisher.subscribe { article -> println("  Subscriber B: $article") }
    publisher.publish("Hello World")
    publisher.publish("Breaking News")
    println("========================================")
}

File: MasteringDart/lib/chapters/chapter17/advanced/advanced_design_patterns.dart
// ignore_for_file: avoid_print
//
//  advanced_design_patterns.dart
//  MasteringDart
//
//  Capítulo 17 — Patrones de diseño.
//  Nivel AVANZADO: Observer — NewsPublisher notifica a suscriptores cuando se publican artículos.
//
//  Ejecutar con: dart run lib/chapters/chapter17/advanced/advanced_design_patterns.dart

typedef Subscriber = void Function(String article);

class NewsPublisher {
  final List<Subscriber> _subscribers = [];

  void subscribe(Subscriber handler) {
    _subscribers.add(handler);
  }

  void publish(String article) {
    for (final sub in _subscribers) {
      sub(article);
    }
  }
}

void runAdvancedDesignPatternExamples() {
  print('========================================');
  print('       OBSERVER PATTERN');
  print('========================================');
  final publisher = NewsPublisher();
  publisher.subscribe((article) => print('  Subscriber A: $article'));
  publisher.subscribe((article) => print('  Subscriber B: $article'));
  publisher.publish('Hello World');
  publisher.publish('Breaking News');
  print('========================================');
}

void main() {
  runAdvancedDesignPatternExamples();
}
